src/lib/route-builder.js

```diff
@@ -0,0 +1,165 @@
+            credits: [{ account: query.destinationAccount, amount: nextHop.finalAmount }]
+          }
+        }
+      }],
+      _hop: nextHop
+    }
+  }
+
+  /**
+   * Given a source transfer with an embedded final transfer, get the next
+   * transfer in the chain.
+   *
+   * @param {Transfer} sourceTransfer
+   * @returns {Transfer} destinationTransfer
+   */
+  * getDestinationTransfer (sourceTransfer) {
+    const traderCredit = sourceTransfer.credits.find(this._isTraderFunds, this)
+    const finalTransfer = traderCredit.memo && traderCredit.memo.destination_transfer
+    if (!finalTransfer) {
+      throw new Error('source transfer is missing destination_transfer in memo')
+    }
+
+    const sourceLedger = sourceTransfer.ledger
+    const finalLedger = finalTransfer.ledger
+    const _nextHop = this.routingTables.findBestHopForSourceAmount(
+      sourceLedger, finalLedger, traderCredit.amount)
+    if (!_nextHop) throwAssetsNotTradedError()
+    const nextHop = yield this._roundHop(_nextHop)
+
+    // Check if this connector can authorize the final transfer.
+    if (nextHop.destinationLedger === finalLedger) {
+      const traderDebit = finalTransfer.debits.find((debit) => !debit.account)
+      if (traderDebit) {
+        traderDebit.account = nextHop.destinationDebitAccount
+      }
+      return finalTransfer
+    }
+
+    return _.omitBy({
+      id: nextHop.destinationLedger + '/transfers/' + uuid(),
+      ledger: nextHop.destinationLedger,
+      debits: [{
+        account: nextHop.destinationDebitAccount,
+        amount: nextHop.destinationAmount
+      }],
+      credits: [{
+        account: nextHop.destinationCreditAccount,
+        amount: nextHop.destinationAmount,
+        memo: {destination_transfer: finalTransfer}
+      }],
+      execution_condition: sourceTransfer.execution_condition,
+      cancellation_condition: sourceTransfer.cancellation_condition,
+      expires_at: this._getDestinationExpiry(sourceTransfer.expires_at)
+    }, _.isUndefined)
+  }
+
+  _findNextHop (query) {
+    return query.sourceAmount
+      ? this.routingTables.findBestHopForSourceAmount(
+          query.sourceLedger, query.destinationLedger, query.sourceAmount)
+      : this.routingTables.findBestHopForDestinationAmount(
+          query.sourceLedger, query.destinationLedger, query.destinationAmount)
+  }
+
+  _getDestinationExpiry (sourceExpiry) {
+    if (!sourceExpiry) return
+    const sourceExpiryTime = (new Date(sourceExpiry)).getTime()
+    const minMessageWindow = this.minMessageWindow * 1000
+    return (new Date(sourceExpiryTime - minMessageWindow)).toISOString()
+  }
+
+  * _roundHop (hop) {
+    hop.sourceAmount = yield this._roundAmount('source', hop.sourceLedger, hop.sourceAmount)
+    hop.destinationAmount = yield this._roundAmount('destination', hop.destinationLedger, hop.destinationAmount)
+    hop.finalAmount = yield this._roundAmount('destination', hop.finalLedger, hop.finalAmount)
+    return hop
```

Comment:
That function is called with the result of `_findNextHop`, not the query. All of the amounts are strings at this point.